ddtrace/internal/processor/trace.py

```diff
@@ -79,9 +82,18 @@ def process_trace(self, trace):
 
                     return single_spans or None
 
+            at_least_one_sampled = False
             for span in trace:
+                # DEV: if the whole trace is being returned, single-span-sampling metrics don't matter anymore
+                if is_single_span_sampled(span):
+                    span._remove_metric(_SINGLE_SPAN_SAMPLING_MECHANISM)
+                    span._remove_metric(_SINGLE_SPAN_SAMPLING_RATE)
+                    span._remove_metric(_SINGLE_SPAN_SAMPLING_MAX_PER_SEC)
                 if span.sampled:
```

Comment:
this is a digression, I wonder if we can safely say... if at least one span is *not* sampled, then the whole trace isn't sampled?

vs the way we have it now is if at least one span is sampled then the trace should be sampled?

I am wondering if we can add a short circuit in here for a case when the trace wouldn't be kept?

(I don't think so, but asking the question anyways)

Reply:
I think this would work when the only sampler in use is the default sampler with its default `SamplingRule`, because it uses the [trace ID](https://github.com/DataDog/dd-trace-py/blob/5388335cfbe3c6d6f4a9ec914eb9422117bca5fc/ddtrace/sampler.py#L504) to make the sampling decision, meaning that `span.sampled` will be the same for all spans in the trace. However, I'm not sure we can maintain that guarantee when other `SamplingRules` are in use.

This question also doesn't seem to affect the performance of the change in this PR, because the single-span-sampling tag check needs to happen on all spans regardless of the state of `span.sampled`, meaning that a full iteration is still necessary.

As I see this currently there's no change necessary to this diff block. What am I missing? Thanks for your input.

Reply:
You are right that it should be the same for all spans, but because it is a mutable property on spans, they could be different.


```python
>>> import ddtrace
>>> with ddtrace.tracer.trace("root") as root:
...     with ddtrace.tracer.trace("child") as child:
...             pass
...
>>> root.sampled
True
>>> child.sampled
True
>>> child.sampled = False
>>> root.sampled
True
>>> child.sampled
False
```

they _should_ be the same, but there is no guarantee :/

Reply:
maybe we don't need to make a change, maybe more "FYI, should we care?"